### PR TITLE
KOGITO-3964 Add RequestHandler Generation for RuleUnits

### PR DESCRIPTION
**Pull request by snowe2010**
In order to call kogito code in AWS, you must use a REST API, therefore
you cannot call directly from lambda to lambda, but must instead go out
through API Gateway, necessitating setting up large amounts of
infrastructure for a simple call from one lambda to another.

Instead, generate RequestHandlers that can create lambda functions that
can be called directly. Copy the interface of the existing Query
Endpoints, but instead implement the RequestHandler interface and
provide a handleRequest method that AWS can use to generate the lambda
functions.

https://issues.redhat.com/browse/KOGITO-3964
https://kie.zulipchat.com/#narrow/stream/232676-kogito/topic/kogito.20handler.20which.20implements.20RequestHandler

- [x] You have read the [contributors guide](CONTRIBUTING.md)
- [x] Pull Request title is properly formatted: `KOGITO-XYZ Subject`
- [x] Pull Request title contains the target branch if not targeting master: `[0.9.x] KOGITO-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

### CODE REVIEW

**Comment by danielezonca** [comment]
ok to test

**Comment by snowe2010** [comment]
The RequestHandler generates just fine when using `kogito:scaffold` but not when using `quarkus:dev`. Not sure what I missed.

**Review by danielezonca — COMMENTED** [review]
Here is an initial quick review

**Comment by danielezonca** [comment]
@snowe2010 
You need to update this class to make Quarkus integration work 
https://github.com/kiegroup/kogito-runtimes/blob/master/kogito-quarkus-extension/deployment/src/main/java/org/kie/kogito/quarkus/deployment/KogitoAssetsProcessor.java
Spring integration is based on Maven plugin while with Quarkus there is a different integration point

File: kogito-codegen/src/main/java/org/kie/kogito/codegen/rules/IncrementalRuleCodegen.java [hunk 1]
```diff
@@ -15,23 +15,6 @@
 
 package org.kie.kogito.codegen.rules;
 
-import java.io.File;
```
[the remark this answers is not in the document]
Comment (snowe2010) [reply kogito-codegen/src/main/java/org/kie/kogito/codegen/rules/IncrementalRuleCodegen.java:18]: Thanks! I looked for that in the CONTRIBUTING.md guide but didn't see any code style, so just went ahead knowing someone would provide it if they cared 😂

File: kogito-codegen/src/main/java/org/kie/kogito/codegen/rules/IncrementalRuleCodegen.java [hunk 2]
```diff
@@ -353,6 +340,21 @@ private void generateRuleUnits( List<DroolsError> errors, List<org.kie.kogito.co
...
+    private List<String> generateHandlers( List<DroolsError> errors, List<org.kie.kogito.codegen.GeneratedFile> generatedFiles,
+                                    RuleUnitHelper ruleUnitHelper, RuleUnitGenerator ruleUnit ) {
+        List<QueryRequestHandlerGenerator> queries = ruleUnit.queriesAsRequests();
+        if (queries.isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        if (annotator == null) {
+            generatedFiles.add( new RuleUnitDTOSourceClass( ruleUnit.getRuleUnitDescription(), ruleUnitHelper ).generateFile( org.kie.kogito.codegen.GeneratedFile.Type.DTO) );
+        }
```
[the remark this answers is not in the document]
Comment (snowe2010) [reply kogito-codegen/src/main/java/org/kie/kogito/codegen/rules/IncrementalRuleCodegen.java]: Looks like moving this out actually results in a subtle bug (or maybe uncovers and existing bug). We only generate the RuleUnitDTO if `ruleUnit.queries()` returns something and if `useRestQueries` is turned on. If `ruleUnit.queries()` doesn't return anything then we never generate the `RuleUnitDTOSourceClass`. 

Is this the _correct_ functionality? Seems weird to me that we'd only generate these if there were some queries. If it is the correct functionality then [this commit](https://github.com/kiegroup/kogito-runtimes/pull/920/commits/5cff7a919ec04148d9a39200554e67fc02396d54) should work. Else I can just use [this commit](https://github.com/kiegroup/kogito-runtimes/pull/920/commits/9da03eb249f92983f810e2fe0579cecddcf3ac2c).

[the remark this answers is not in the document]
Comment (evacchi) [reply kogito-codegen/src/main/java/org/kie/kogito/codegen/rules/IncrementalRuleCodegen.java]: /cc @mariofusco who added some of these checks/features

[the remark this answers is not in the document]
Comment (mariofusco) [reply kogito-codegen/src/main/java/org/kie/kogito/codegen/rules/IncrementalRuleCodegen.java]: > Is this the correct functionality?
Yes, it is supposed to work that way. That DTO for the rule unit is necessary to marshall the unit itself when invoking the rest endpoint generated for that query, For this reason it should be generated only if there is at least one query and if generation of automatic rest endpoints is turned on. In all other cases generating it won't be wrong, but just wasteful.

[the remark this answers is not in the document]
Comment (snowe2010) [reply kogito-codegen/src/main/java/org/kie/kogito/codegen/rules/IncrementalRuleCodegen.java]: @mariofusco if I'm generating RequestHandlers that use the same dto marshalling then I would need to generate if either rest queries were enabled or request handlers were enabled. Correct?

**Comment by natedennis** [comment]
did we give up on this? i sort of need this functionality

**Comment by snowe2010** [comment]
@natedennis It's not been given up, but I can't get any feedback from the team in regards to why Quarkus will not find the RequestHandlers. Here are the relevant conversations:

* https://kie.zulipchat.com/#narrow/stream/232676-kogito/topic/kogito.20handler.20which.20implements.20RequestHandler/near/220317801
* https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/AWS.20RequestHandlers.20not.20detected.20if.20generated/near/220408883

**Comment by danielezonca** [comment]
Hi @snowe2010 
Sorry for delay but after end of year I didn't have time to look at the PR.
Today I updated your PR with latest changes from master (we changed many things in `kogito-codegen` module and other will come).
I also refactored `QueryEndointGenerator` to be extensible so that `QueryRequestHandlerGenerator` can extend it and reuse most of the logic.
Finally I moved the logic to have REST or `RequestHandler` in a single method inside `RuleUnitGenerator.toQueryEndpointGenerator`. See this commit https://github.com/snowe2010/kogito-runtimes/pull/1/commits/43f5e936f127bc0d83cb4b2b9ddf7f47d2c9ae63 to check only most of the changes

Here is the PR to your branch
https://github.com/snowe2010/kogito-runtimes/pull/1

As far as I can see (using your example https://github.com/kiegroup/kogito-examples/pull/484 ) now the file are generated correctly (check `generated-sources/kogito`).
It is also necessary to manually add a property [`quarkus.lambda.handler`](https://quarkus.io/guides/amazon-lambda#choose) with the name of the query because there are two of them (by default the extension expects to find a single impl of `RequestHandler`).

Unfortunately it is still not working because of the way `quarkus-amazon-lambda-deployment` introspect the project to find the lambdas: it is not considering `GeneratedBeanBuildItem`s (bean produced during codegen) but only a specific set of types ([see list](https://github.com/quarkusio/quarkus/blob/master/extensions/amazon-lambda/deployment/src/main/java/io/quarkus/amazon/lambda/deployment/AmazonLambdaProcessor.java#L182-L188)).

The reason why it is working with `scaffold` mojo is because the code is actually copied inside a source folder (by the plugin) so then the quarkus extension can find the code because it has been already generated by the previous step.

I need to better check the behavior of that quarkus extension component and see how to make it more flexible. In the worst case scenario it is possible to define an hoc `kogito-quarkus-lambda` extension but this will take more time and it is not ideal in general.

I hope to have other updates soon

**Comment by snowe2010** [comment]
@danielezonca thank you for the hard work. 

>It is also necessary to manually add a property quarkus.lambda.handler with the name of the query because there are two of them (by default the extension expects to find a single impl of RequestHandler).

Yes I was wondering how to accomplish that, due to the query generating two. Glad you have a solution for that. 

>Unfortunately it is still not working because of the way quarkus-amazon-lambda-deployment introspect the project to find the lambdas: it is not considering GeneratedBeanBuildItems (bean produced during codegen) but only a specific set of types (see list).

What part of that is limiting the types? Not sure I see the list. 

>The reason why it is working with scaffold mojo is because the code is actually copied inside a source folder (by the plugin) so then the quarkus extension can find the code because it has been already generated by the previous step.

Makes sense. 

>I need to better check the behavior of that quarkus extension component and see how to make it more flexible. In the worst case scenario it is possible to define an hoc kogito-quarkus-lambda extension but this will take more time and it is not ideal in general.

> I hope to have other updates soon

Awesome. If there is something I can look into from here let me know. It seems like the solution is to make `quarkus-amazon-lambda-deployment` consider those beans generated during codegen, but I'm guessing there's a reason why they didn't do that.

**Comment by snowe2010** [comment]
@danielezonca Hi Daniel, I'm wondering if you had any updates from your side on this? It's getting quite difficult to create aws handlers, though I'm wondering if this PR will end up actually being a solution for us. We're switching everything over to kotlinx-serialization, and as a result `RequestHandler`s don't actually work for us anymore, due to jackson serialization. We're using `RequestStreamHandler`s instead, and serializing in the handler itself.

**Comment by danielezonca** [comment]
> @danielezonca Hi Daniel, I'm wondering if you had any updates from your side on this? It's getting quite difficult to create aws handlers, though I'm wondering if this PR will end up actually being a solution for us. We're switching everything over to kotlinx-serialization, and as a result `RequestHandler`s don't actually work for us anymore, due to jackson serialization. We're using `RequestStreamHandler`s instead, and serializing in the handler itself.

Hi @snowe2010,

I was on holiday last week :)

I think this feature it is still in the backlog and I'm having no time to work on it so it is on hold at the moment.
Can you share a small project with the `RequestStreamHandler` like you are using it to see if it is something we can consider?
What is in particular the limitation you are getting with `RequestHandler`? 
As far as I see `RequestHandler` has been deprecated by AWS so the current approach will not work in any case so if you need something different we can consider it

**Comment by snowe2010** [comment]
>Hi @snowe2010,

>I was on holiday last week :)

>I think this feature it is still in the backlog and I'm having no time to work on it so it is on hold at the moment.
Can you share a small project with the RequestStreamHandler like you are using it to see if it is something we can consider?
What is in particular the limitation you are getting with RequestHandler?
As far as I see RequestHandler has been deprecated by AWS so the current approach will not work in any case so if you need something different we can consider it


Hi Daniel, no worries. I have a project in a bug I filed [here](https://issues.redhat.com/browse/KOGITO-5523) that demonstrates a different issue, but it contains the RequestStreamHandler and example of how we are using it. It also demonstrates a different issue, in that due to using kogito:scaffold, we're now unable to build the project without `mvn clean`ing every time. 

The main issue with `RequestHandler` is that because kogito/quarkus have hard dependencies on jackson, we're unable to decouple the serialization that is done in the `RequestHandler` from it and use `kotlinx-serialization` instead. We've encountered many issues with Jackson in Graal, so we are trying to switch over to kotlinx-serialization which doesn't have these issues. Therefore, to avoid using jackson for serialization in the handler, we have to use the stream handler, rather than the regular one. I hadn't noticed that `RequestHandler` was deprecated, so that is news to me, so I guess it makes sense to not support that anyway.

**Comment by ricardozanini** [comment]
@snowe2010 can you rebase?

**Comment by snowe2010** [comment]
@ricardozanini Looks like the codebase completely changed from when I first created the PR. It will probably not be possible to rebase. The work will have to be done over again.

**Comment by domhanak** [comment]
Closing as stale/inactive. Please reopen if required.

**Pull request closed without merging**